working/0546-patterns/goals-and-constraints.md

```diff
@@ -0,0 +1,461 @@
+lexical scope resolution to determine if a pattern is a constant pattern or a
+variable pattern.
+
+We'll presumably want to support dotted identifiers for enum cases. I imagine
+users will expect that to also work for named constants imported from libraries
+with a prefix.
+
+That raises the question of whether we want to support arbitrary constant
+expressions. That has some appeal, but means that much of the expression grammar
+would overlap the pattern grammar, which leads to ambiguity and pain. Consider,
+for example, `[1, 2]`. Is that a constant pattern containing a constant list
+literal, or a list destructuring pattern containing two literal patterns?
+
+### Destructure collections
+
+A key goal and a feature offered by other languages is patterns that can match
+and pull elements out of collections. Dart already has lists, maps, and sets.
+Patterns to let you destructure those are an obvious feature.
+
+We also want to add [tuples][] and those almost require destructuring because it's
+difficult to otherwise access the elements in a type safe way, since tuples are
+heterogeneously typed.
+
+[tuples]: https://github.com/dart-lang/language/issues/68
+
+As in other languages, the syntax for these should mirror the corresponding
+collection literal expression form:
+
+```dart
+Object obj = ...
+match (obj) {
+  case [a, b] => print("list with elements $a and $b");
+  case {a, b} => print("set with elements $a and $b");
+  case (a, b) => print("tuple with elements $a and $b");
+}
+```
+
+Maps are a little trickier. Do they match specific entries, in which case the
+key is a value, or do they destructure any entry, in which case the key is a
+pattern?
+
+### Destructure instances of classes
+
+Another common use of pattern matching is for working with [sum types][]. In
+typed functional languages, sum types are the typical way to define values that
+can be one of a few different cases, with potentially nested data for each case.
+Pattern matching is then how you implement behavior that varies based on which
+case you have.
+
+[sum types]: https://en.wikipedia.org/wiki/Tagged_union
+
+In an object-oriented language, those cases are modeled using subclasses. Thus,
+to use pattern matching to select behavior based on a sum type case, we need
+patterns that match a specific subclass and extract fields from it.
+
+For example, something like this:
+
+```dart
+abstract class Shape {}
+class Rect extends Shape {
+  final num left, top, right, bottom;
+  Rect(this.left, this.top, this.right, this.bottom);
+}
+class Circle extends Shape {
+  final num x, y, radius;
+  Circle(this.x, this.y, this.radius);
+}
+
+printShape(Shape shape) {
+  match (shape) {
+    case Rect(left, top, right, bottom) =>
+        print("Rect $left, $top, $right, $bottom");
+    case Circle(x, y, radius) =>
+        print("Circle $x, $y, $radius");
+  }
+}
+```
+
+Here, the `Rect(...)` and `Circle(...)` are matching on shapes of the
```

Comment:
There are interesting design choices here around what this is testing.  Exact instance of Rect? Subclass of Rect? Instance of something implementing Rect?  Instance of something conforming to the "Rect destructor protocol"?  etc.

Reply:
Yes, or even an arbitrary class that happens to implement some destructuring protocol that accepts a Shape as the matched value.